kagami/routes/admin/resource.py

```diff
@@ -1,18 +1,34 @@
-from fastapi import APIRouter, HTTPException
-from pydantic import BaseModel
-from ...deps import SupervisorDeps
+
 
 from fastapi import APIRouter, HTTPException
-from pydantic import BaseModel
-from ...deps import SupervisorDeps
+from pydantic import BaseModel, HttpUrl, field_validator
+from typing import Literal
+
+from kagami.routes.deps import SupervisorDeps
 
 resource_router = APIRouter(prefix="/admin/resource")
 
 class ResourceConfig(BaseModel):
     name: str
-    upstream_url: str
-    provider_method: str
+    upstream_url: HttpUrl
+    provider_method: Literal["rsync", "git"]
```

Comment:
There is no need to change str into HttpUrl or Literal["rsync", git].
Upstream URL can be other protocol and HttpUrl may cause pydantic confused.
Type hint does not check, and provider method is defined by Provider. If we introduce plugin methods to add more provider_method, user should update this code every time when a new protocol is introduced.